Add additional test coverage for helper functions in Util

gajananan opened the pull request:
# Summary

Currently helper functions in Util lacks test coverage

This commit adds test coverage helper functions in Util

Ref #4380

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [x] Checked that related changes are merged.

JAORMX commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+			t.Parallel()
+
+			// Create a unique directory for each test
+			testDir := fmt.Sprintf("testdir_%s", sanitizeTestName(tt.name))
```
You could probably leverage golang's utilities instead: https://pkg.go.dev/testing#B.TempDir

This will create a temporary directory for tests and even clean up afterwards.

wdyt?

gajananan replied:
thank you. I agree. that is a clean approach.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+
+			// Create a temporary file for testing
+			if tc.filePath == "testfile.txt" {
+				err := os.WriteFile(tc.filePath, []byte(tc.expectedDesc), 0600)
```
We should do this in `t.TestDir()`

gajananan replied:
Thank you. I assume, you were referring to `t.TempDir()`

evankanderson replied:
Yes, thanks!

evankanderson replied:
Alternatively, you could create `testfile.txt` once outside the `t.Run()`, since it shouldn't need to vary across the test cases.

evankanderson replied:
(You probably still want to put it in `t.TempDir()`, though.)

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+// TestRemoveCredentials tests the RemoveCredentials function
+func TestRemoveCredentials(t *testing.T) {
+	t.Parallel()
+	xdgConfigHome := os.Getenv("XDG_CONFIG_HOME")
+	if xdgConfigHome == "" {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			t.Fatalf("error getting home directory: %v", err)
+		}
+		xdgConfigHome = filepath.Join(homeDir, ".config")
+	}
+
+	filePath := filepath.Join(xdgConfigHome, "minder", "credentials.json")
```
Thank you. it is good to learn about the impact to other tests.

evankanderson commented:
@gajananan

It looks like this is close to the finish line -- any chance you can incorporate the feedback (and fix conflicts) in the next week?

evankanderson commented:
Hi @gajananan,

We'd love to get this merged!  If you don't have time to work on this, we'll probably have someone over here at Stacklok apply the requested changes and merge it.

evankanderson commented:
Hey @gajananan -- we'd love to get this PR in.  I ran into one of your co-workers at KubeCon, and they said that your work schedule might be preventing you from updating.  A quick note of "I can get to this within a week" or "please carry this forward, I'm busy now" would be handy.

Thanks again for your work so far!

gajananan commented:
> Hey @gajananan -- we'd love to get this PR in. I ran into one of your co-workers at KubeCon, and they said that your work schedule might be preventing you from updating. A quick note of "I can get to this within a week" or "please carry this forward, I'm busy now" would be handy.
> 
> Thanks again for your work so far!

My sincere apologies for long silence on this.  I would take it forward with addressing the comments/feedback

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			err := os.Setenv("XDG_CONFIG_HOME", tt.envVar)
```
I will add explicit locks.

evankanderson replied:
(This still needs to clean up after the `os.Setenv`)

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			err := os.Setenv(MinderAuthTokenEnvVar, tt.envToken)
```
I will do reset the env at the end of test run.

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+			parsedURL, _ := url.Parse(server.URL)
+			tt.issuerUrl = parsedURL.String()
+
+			result, err := RefreshCredentials(tt.refreshToken, tt.issuerUrl, tt.clientId)
```
@evankanderson , could you confirm we still need to do this after my latest change?

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+	}
+
+	for _, tt := range tests {
+		tt := tt // capture range variable
```
I think this is no longer needed [since Go 1.22](https://forum.golangbridge.org/t/go-1-22-changed-for-loop-variable-semantics/34464).

gajananan replied:
Yes, no longer need.  I removed it.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+		tt := tt // capture range variable
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			// Create a unique temporary directory for the test
+			// tempDir, err := os.MkdirTemp("", "test_load_credentials_"+tt.name)
+			// if err != nil {
+			// 	t.Fatalf("failed to create temp directory: %v", err)
+			// }
+			// defer os.RemoveAll(tempDir)
```
Why is this commented out?

gajananan replied:
It should be removed.  I will do that.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -13,21 +13,765 @@
+			}
+
+			// Temporarily override the environment variable for the test
+			originalEnv := os.Getenv("XDG_CONFIG_HOME")
+			err = os.Setenv("XDG_CONFIG_HOME", tempDir)
```
This still doesn't have a `defer` or switched to `t.Setenv`.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+					t.Errorf("expected result %v, got %v", tt.expectedResult, result)
+				}
+			}
+			defer os.Setenv("XDG_CONFIG_HOME", originalEnv)
+			defer os.RemoveAll(tempDir)
```
If you're using `defer` (instead of `t.Cleanup`), you should put it next to the resource it creates. By putting these at the end, you'll only perform the cleanup if the test doesn't exit early (e.g. from `t.Error`)

gajananan replied:
I moved the deferred call as suggested.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			var server *httptest.Server
+			if tt.name != "Invalid issuer URL" {
```
Rather than using the test name, why not have a `func(t *testing.T) *httptest.Server` in the test cases which constructs the server?

gajananan replied:
I modified to have a func(t *testing.T) *httptest.Server in the test cases which constructs the server.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+			var server *httptest.Server
+			if tt.name != "Invalid issuer URL" {
+				server = httptest.NewServer(http.HandlerFunc(func(_ http.ResponseWriter, r *http.Request) {
+					if r.URL.Path != tt.expectedPath {
+						t.Errorf("expected path %s, got %s", tt.expectedPath, r.URL.Path)
+					}
```
We use [require](https://pkg.go.dev/github.com/stretchr/testify/require) in other parts of the codebase to write this as:

```golang
require.Equal(t, tt.expectedPath, r.URL.Path)
```

You're not required to do so, but it (and `NoError`) can be helpful sometimes.

gajananan replied:
That is good to learn. I changed accordingly.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+						t.Errorf("expected token_type_hint %s, got %s", tt.tokenHint, r.Form.Get("token_type_hint"))
+					}
+				}))
+				defer server.Close()
```
You can't put this inside a constructor function if you take that approach.

gajananan replied:
I addressed this as I have changed the approach with test server creation.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+				defer server.Close()
+			}
+
+			issuerUrl := tt.issuerUrl
+			if tt.name != "Invalid issuer URL" {
+				issuerUrl = server.URL
+			}
```
Since most of the body of this test is conditional on the test name, I wonder if using two tests (rather than a data-driven test) would simplify things here.

gajananan replied:
I significantly  changed the test function, I guess this comment has been addressed with latest change.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				if tc.filePath == "-" || tc.filePath != "-" {
```
Doesn't this `if` cover all cases?

gajananan replied:
The condition if tc.expectError { assert.Error(t, err) } covers only the test cases where an error is expected.

The test cases with no expected error ( there are two)  are covered by the subsequent else block, which checks that no error occurred and verifies the file content.

gajananan replied:
Else condition covers the case:
	    {
			name:         "Dash as file path",
			filePath:     "-",
			dashOpen:     strings.NewReader("dash input"),
			expectedDesc: "dash input",
			expectError:  false,
		}

gajananan replied:
"Alternatively, you could create testfile.txt once outside the t.Run()"

 testfile.txt  is required only in one case, so it is created only for that case.

evankanderson replied:
Sorry, the comment here is that this line says:

```suggestion
				path := tc.filePath
				if path == "-" || path != "-" {
```

It should always be true that a string is either equal to a constant, or not equal to the same constant.

evankanderson replied:
> "Alternatively, you could create testfile.txt once outside the t.Run()"
> 
> testfile.txt is required only in one case, so it is created only for that case.

(This was from a different comment.)

While it's true that `testfile.txt` is only needed in one case, it doesn't hurt the other cases, and could end up simplifying the setup logic.  On the other hand, I don't feel strongly about it being one place or the other.

gajananan replied:
> Sorry, the comment here is that this line says:
> 
> It should always be true that a string is either equal to a constant, or not equal to the same constant.
I got it, I will fix it.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+	}
+}
+
+func compareExpandedFiles(got, expected []util.ExpandedFile) bool {
+	if len(got) != len(expected) {
+		return false
+	}
+	for i := range got {
+		if got[i].Path != expected[i].Path || got[i].Expanded != expected[i].Expanded {
+			return false
+		}
+	}
+	return true
+}
```
Consider using [`slices.Equal`](https://pkg.go.dev/slices#Equal), or `slices.EqualFunc` with an anonymous function if you need to.

gajananan replied:
I will use slices.EqualFunc with an anonymous function

evankanderson reviewed (commented):
Thanks for getting back to this.  Some of the comments are nit-picky, and you can add a comment saying "just trying to get this in", but since I thought part of your goal was learning the Go language, I figured I'd provide more extensive comments and pointers.

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+	// Create a temporary directory
+	testDir := t.TempDir()
+
+	err := os.Setenv("XDG_CONFIG_HOME", testDir)
```
I will reset once the test is completed.

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,767 @@
+					t.Errorf("error closing connection: %v", err)
+				}
+			}
+			defer os.Setenv(util.MinderAuthTokenEnvVar, originalEnvToken)
```
I moved this deferred call up next to the os.Setenv call.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+// TestGetConfigDirPath tests the GetConfigDirPath function
+func TestGetConfigDirPath(t *testing.T) {
+	t.Parallel()
+	var envLock sync.Mutex
```
I think you want `envLock` to be at the file level, e.g. on line 32 before this function, because you have multiple tests all trying to set the environment, and you don't want any of them conflicting with each other.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			originalEnvToken := os.Getenv(util.MinderAuthTokenEnvVar)
+			err := os.Setenv(util.MinderAuthTokenEnvVar, tt.envToken)
+			// reset this the environment variable when complete.
+			defer os.Setenv(util.MinderAuthTokenEnvVar, originalEnvToken)
```
You'll want to protect this with a mutex as well.  Given the likely test execution time, I'd probably be lazy and re-use `envLock` from `TestGetConfigDirPath`,  but you could also use separate locks for each environment variable if you're feeling particularly ambitious (i wouldn't recommend it, though).

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+// TestSaveCredentials tests the SaveCredentials function
+func TestSaveCredentials(t *testing.T) {
+	t.Parallel()
+	var envLock sync.Mutex
+	envLock.Lock()
+	defer envLock.Unlock()
```
Declaring a new `sync.Mutex` here and then acquiring it doesn't do much, because the mutex isn't shared across multiple goroutines (e.g. `t.Run()` spawns a new goroutine, but this doesn't call `t.Run`).

gajananan replied:
Fixed now.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			tempDir := t.TempDir()
+			defer os.RemoveAll(tempDir)
```
You shouldn't need this if you use [`t.TempDir()`](https://pkg.go.dev/testing#T.TempDir):

> The directory is automatically removed when the test and all its subtests complete.

gajananan replied:
Removed

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+				t.Logf("Test %s: file path %s not created as file content is empty", tt.name, filePath)
+			}
+
+			// Temporarily override the environment variable for the test
+			originalEnv := os.Getenv("XDG_CONFIG_HOME")
+			err = os.Setenv("XDG_CONFIG_HOME", tempDir)
+			if err != nil {
+				t.Errorf("error setting XDG_CONFIG_HOME: %v", err)
+			}
+			// reset this the environment variable when complete.
+			defer os.Setenv("XDG_CONFIG_HOME", originalEnv)
```
You'll want to grab `envLock` here for the rest of the test, too.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+	createServer func(t *testing.T, tt TestCase) *httptest.Server
+}
+
+func createTestServer(t *testing.T, tt TestCase) *httptest.Server {
```
golangci-lint likes non-top-level test functions that take a `t *testing.T` to call `t.Helper()` so that the tracing of errors from `t.Errorf()` etc gets a cleaner stacktrace.

gajananan replied:
t.Helper() is added to createTestServer().

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+				tempFilePath = filepath.Join(testDir, tc.filePath)
+				err := os.WriteFile(tempFilePath, []byte(tc.expectedDesc), 0600)
+				assert.NoError(t, err)
+				defer os.Remove(tempFilePath)
```
Since you're putting `tempFilePath` in a directory from `t.TempDir`, you shouldn't need to delete it yourself later; `t.TempDir` will delete the directory and its contents at the end of the test.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,758 @@
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			// Create a temporary file for testing
+			if tc.filePath == "testfile.txt" {
+				tempFilePath = filepath.Join(testDir, tc.filePath)
+				err := os.WriteFile(tempFilePath, []byte(tc.expectedDesc), 0600)
+				assert.NoError(t, err)
+				defer os.Remove(tempFilePath)
+			} else {
+				// When handling the dash (-) as a file path, it should read from the provided io.Reader (tc.dashOpen) instead of trying to open a file
+				tempFilePath = tc.filePath
+			}
```
If you create `testfile.txt` outside the test case code, you can avoid needing as much conditional logic in the test.  Compare:

```golang
func TestOpenFileArg(t *testing.T) {
	t.Parallel()
	testDir := t.TempDir()
	// Create one test file for the file input case
	tempFilePath := filepath.Join(testDir, "testfile.txt")
	assert.NoError(os.WriteFile(tempFilePath, []byte("test content"), 0600))

	...
		filePath: tempFilePath
```

(vs the current 11 lines with `if`/`else`)

gajananan replied:
thank you, I fixed it.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+		t.Fatalf("error marshaling credentials: %v", err)
+	}
+
+	fpath := filepath.Clean(filePath)
```
Why do you need to run `filepath.Clean` here?

gajananan replied:
I tested, we would not need it.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+	expectedFilePath := filepath.Join(cfgPath, "credentials.json")
+
+	filePath, err := util.SaveCredentials(tokens)
+	if err != nil {
+		t.Fatalf("expected no error, got %v", err)
+	}
```
Slight preference for `require.NoError(t, err)` here and below, to break up the common-case flow of the test a bit less.

gajananan replied:
will do that.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+		t.Errorf("expected file content %v, got %v", string(credsJSON), string(content))
+	}
+
+	// Clean up
+	err = os.Remove(filePath)
+	if err != nil {
+		t.Fatalf("error removing file: %v", err)
+	}
```
You don't need this since your files are in `t.TempDir()`; the test will do it for you automatically.

gajananan replied:
Yes, I removed that.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+			}))
+			defer server.Close()
+
+			parsedURL, _ := url.Parse(server.URL)
+			tt.issuerUrl = parsedURL.String()
```
Why do this, rather than simply:

```
tt.issuerUrl = server.URL
```

(or simply use `server.URL` below, and don't make the `issuerUrl` a field in the testcase struct, since it is never used...)

gajananan replied:
yes. I can simplify as suggested.

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+
+			if tt.fileContent != "" {
+				// Create a temporary file with the specified content
+				err := os.WriteFile(filePath, []byte(tt.fileContent), 0600)
+				if err != nil {
+					t.Fatalf("failed to write test file: %v", err)
+				}
```
Again, prefer `require.NoError()`:

```suggestion
				require.NoError(t, os.WriteFile(filePath, []byte(tt.fileContent), 0600))
```

evankanderson commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+			tokenHint:    "refresh_token",
+			expectedPath: "/realms/stacklok/protocol/openid-connect/revoke",
+			expectError:  false,
+			createServer: createTestServer,
```
Another option here is to make createTestServer an anonymous function, which avoids needing to declare it or `TestCase` outside this function:

```suggestion
			createServer: func(t *testing.T) *httptest.Server {
				return httptest.NewServer(http.HandlerFunc(func(_ http.ResponseWriter, r *http.Request) {
					require.NoError(t, r.ParseForm())
					require.Equal(t, "minder-cli", r.Form.Get("client_id"))
					require.Equal(t, "test-token", r.Form.Get("token"))
					require.Equal(t, "refresh_token", r.Form.Get("token_type_hint"))
				}))
			},
```

evankanderson reviewed (approved):
A few more small nits, but this is looking very close!  In particular, I think you want `SetEnvVar` to take care of restoring the environment variable; the other fixes are purely your choice / passing along some experience maintaining Go code.

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+	envLock = &sync.Mutex{}
+)
+
+func SetEnvVar(t *testing.T, env string, value string) {
```
I fix it.

gajananan commented:
> A few more small nits, but this is looking very close! In particular, I think you want `SetEnvVar` to take care of restoring the environment variable; the other fixes are purely your choice / passing along some experience maintaining Go code.

Thank you for your detailed feedback which helped a lot to learn the code base. I addressed the remainder of the comments as well.

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+	envLock.Lock()
+	t.Cleanup(envLock.Unlock)
+
+	err := os.Setenv(env, value)
+	if err != nil {
+		t.Errorf("error setting %v: %v", env, err)
+	}
```
added a cleanup to reset the environment variable to the original value

gajananan commented on internal/util/helpers_test.go:
```diff
@@ -4,18 +4,765 @@
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			originalEnvToken := os.Getenv(util.MinderAuthTokenEnvVar)
+			err := os.Setenv(util.MinderAuthTokenEnvVar, tt.envToken)
```
yes,
